Draft/RFC0001-PowerShell-Based-LCM.md

```diff
@@ -0,0 +1,92 @@
+predominently as a Windows based configuration management tool.
+
+The current implementation is based on WMI/OMI and receives MOF files either from a Pull Server
+or by use of the cmdlets `Start-DscConfiguration` or `Publish-DscConfiguration`.
+
+The LCM can be configured to pull its configuration on a recurring basis and implements security
+by verifying signatures of MOFs or modules before they are downloaded.
+
+The LCM has 3 application modes `ApplyOnly`, `ApplyAndMonitor` and `ApplyAndAutoCorrect`. When
+the configuration is applied, it is applied in a serial fashion i.e. one resource after another.
+There is a dependency model defined within the configuration which enforces the order with which
+resources are deployed.
+
+## Motivation
+
+As a user, I want to replace the existing WMI/OMI based LCM on Windows and Linux with cross
+platform version that can perform all of the tasks of the existing LCM
+plus a lot lot more. So I can utilise my existing DSC knowledge to manage infrastructure
+all over my organisation. Onprem and in the cloud.
+
+I see this being the new Terraform, Polumi etc
+
+## Specification
+
+### User Experience
+
+```PowerShell
+# Command to generate new DSC MOF file
+Build-DscConfiguration
+
+# Command to execute DSC on a machine
+Invoke-DscConfiguration
+```
+
+### The basic Set
+This RFC proposes a way of using PowerShell to perform the task of the LCM:
+
+```PowerShell
+param([Resources[]]$resources)
+
+# Loop each resource and use the Invoke-DscResource cmdlet available in both Windows Powershell
+# and PowerShell Core to start each resource.
+
+foreach($resource in $resources) {
+    $InvokeParameters = @{
+        name = $resource.name
+        module = $resource.module
+        properties = $resource.properties
+    }
+
+    if ($false -eq (Invoke-DscResource @InvokeParameters -Method Test)) {
+        Invoke-DscResource @InvokeParameters -Method Set
+    }
+}
+```
+
+This is a very trivial implementation that just demonstrates how we could implement the set
+functionality of the LCM within PowerShell.
+
+### Dependency model
+
+I recommend we keep the near enough same data model for resources i.e. `resourceid`, `modulename`,
+`moduleversion` and `dependson`. I would make a seperate `properties` object with all the resource
+properties just to seperate the instance data from the meta data.
+
+On initial run, all the resources without any dependencies would get added to a "deployment" queue
+which will be looped over until empty. As resources that are successfully deployed, the dependent
+resources will be discovered and dynamically added to the queue. Any remaining resources will be
+considered to have failed due to missing dependencies.
+
+### Alternate Proposals and Considerations
+
+The RFC is open to comment
+
+#### Open Questions
+
+* Should we bother supporting Windows Powershell?
+* Which file formats should be support? MOF, Json, Yaml
```

Comment:
We should support configuration serialized as MOF using the existing capabilities (supported in Windows PS 5.1, PS7-Preview.4+ and *I think* PS 6.2.1).

I don't think that LCM should support Windows PowerShell (as long as MOF are supported and apply to a consistent result).

I think we should clarify a bit the scope of this effort to start with, and in what order we'll do it.
1. try to provide the "functionalities" of the previous LCM (similar & consistent, not necessarily like for like).
2. Try to make the functionalities available as a service communicating via an API (whether to support old Pull Server is a question we can discuss later)
3. Provide the basics of a central server that can communicate and manage that service

My point is the core functionalities can and should be done in a most generic way, so that third parties (AWS, Chef, Puppet, Ansible & others) can also rely on them should they wish to do so.

The service wrapper is something that would be specific to the solution. Third parties would use their own agent, so it will differ.

At last, the central server should probably integrate other functionalities (by that I mean be extensible), that's clearly out of scope for now.

Reply:
I'm happy to drop WinPS support, so that fine for me.

This is just the RFC pertaining to the engine. There are so many feature I would like to implement here, not only standard API access to provide better integration.

The scope for me is as you've mentioned. I will add it to the PR but just for santities sake. I want to keep as many of the functional features that we have today. i.e. pull server interop, both file and http. the three modes, apply only, apply and monitor and apply and autocorrect. It would be great to have cross platform support with equal feature sets and finally increased performance not only by means of parallelism